In [1]:
## CONVOLUTIONAL NN or CNN
import os
import tensorflow as tf
from tensorflow import keras
import numpy as np
from keras.datasets import mnist

### Regular Neural Network

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [3]:
# Normalize pixel values
x_train = x_train / 255.0
x_test = x_test / 255.0

In [4]:
class my_callback(tf.keras.callbacks.Callback):
    def on_epoch_end(self, epoch, logs={}):
        if(logs.get('accuracy') > 0.99):
            print("\nReached 99% accuracy so cancelling training!")

            self.model.stop_training = True

In [8]:
callback = my_callback()

model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

model.fit(
    x_train,
    y_train,
    epochs=10,
    callbacks=[callback]
)

model.evaluate(x_test, y_test)

Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2499 - accuracy: 0.9295
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1095 - accuracy: 0.9678
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0742 - accuracy: 0.9769
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0540 - accuracy: 0.9841
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0424 - accuracy: 0.9866
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0329 - accuracy: 0.9897
Epoch 7/10
1860/1875 [============================>.] - ETA: 0s - loss: 0.0276 - accuracy: 0.9914
Reached 99% accuracy so cancelling training!
313/313 [==============================] - 1s 1ms/step - loss: 0.0802 - accuracy: 0.9754


[0.08017904311418533, 0.9753999710083008]

### Convolutional Neural Network

In [10]:
callback = my_callback()

cnn_model = keras.Sequential([

    ## Convolutional Portion
    keras.layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    keras.layers.MaxPooling2D(2, 2),
    keras.layers.Conv2D(64, (3, 3), activation='relu'),
    keras.layers.MaxPooling2D(2, 2),

    ## Basic Neural Network Portion
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

cnn_model.summary()

cnn_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

print('\nTraining Model')
cnn_model.fit(
    x_train,
    y_train,
    epochs=10,
    callbacks=[callback]
)

print('\nEvaluation Score')
cnn_model.evaluate(x_test, y_test)

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 64)       0         
 2D)                                                             
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 64)        36928     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 flatten_3 (Flatten)         (None, 1600)              0         
                                                                 
 dense_6 (Dense)             (None, 128)              

[0.030563855543732643, 0.9890999794006348]

As we can see, using the convolutional filters on top of our original DNN, we are able to acheive a much better Loss, Accuracy score.

DNN:

- Loss: .08
- Accuracy: .97

CNN:

- Loss: .03
- Accuracy: .989